### PR TITLE
Epics feature

### PR DESCRIPTION
**Pull request by 88mary256**

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 app:
  #options in level
- token: ""
+ token: "290548f427d7f7d10b461e95036ea3a3"
```
Comment (d4vidmc) [inline config.yml]: delete this

File: config.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
...
  base_url: "https://www.pivotaltracker.com/services/v5"
- logger_path:
\ No newline at end of file
+ logger_path: "\\resources\\logs"
```
Comment (d4vidmc) [inline config.yml]: change to \log

File: features/epics.feature [hunk 3]
```diff
@@ -0,0 +1,38 @@
+Feature: Epics
+
+  Scenario: Get All epics of a project
+    Given I make a connection to epics
+    When I get all epics of project with id 2322921
```
Comment (d4vidmc) [inline features/epics.feature]: do not hardcode the id of an epic

File: features/epics.feature [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    Then I validate epic list is retrieved
+
+
+  Scenario: Get all epics with filter
+    Given I make a connection to epics
+    When I get all epics of project with id 2322921 and filter {state:unstarted}
```
Comment (d4vidmc) [inline features/epics.feature]: Do not hardcode the project id

File: features/epics.feature [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    Then I validate epic list is retrieved
+
+  Scenario: Get 1 epic with id from project
+    Given I make a connection to epics
+    When I get epic with id 164996934 from project 2322921
```
Comment (d4vidmc) [inline features/epics.feature]: Do not hardcode the id of the project and epic

File: features/steps/epics_steps.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+@when("I get all epics of project with id {id} and filter {filter}")
+def step_get_all_epic_of_project(context, id, filter):
+    context.all_epics_response = context.service.get_all_epics_with_filter(id, filter)
+
+
+@then("I validate epic list is retrieved")
+def step_validate_epics_list(context):
+    expect(context.all_epics_response.status_code)== 200
+    epics = context.all_epics_response.json()
+    pprint.pprint(epics)
```
Comment (d4vidmc) [inline features/steps/epics_steps.py]: use logger instead of print statement

File: features/steps/epics_steps.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+@when('I modify name of epic with id {epic_id} in project {project_id} to "{name}"')
+def step_modify_epic(context, epic_id, project_id, name):
+    context.epic_response = context.service.modify_epic(project_id, epic_id, "{'name':'"+name+"'}")
+
+@then("I validate epic result")
+def step_validate_epic(context):
+    expect(context.epic_response.status_code) == 200
+    epic = context.epic_response.json()
+    pprint.pprint(epic)
+    expect(epic) != ""
```
Comment (d4vidmc) [inline features/steps/epics_steps.py]: where are the error messages

File: features/steps/project_steps.py [hunk 8]
```diff
@@ -1,25 +1,26 @@
 import logger as logger
 
 from behave import given, when, then
-from pip import logger
+#from pip import logger
```
Comment (d4vidmc) [inline features/steps/project_steps.py]: remove this

File: features/steps/project_steps.py [hunk 9]
```diff
@@ -1,25 +1,26 @@
...
 
 from src.pivotal_services.project_service import ProjectService
 
 
 @given(u'I make a connection')
 def step_impl(context):
-    logger.info(u'STEP: Given I make a connection')
+#    logger.info(u'STEP: Given I make a connection')
```
Comment (d4vidmc) [inline features/steps/project_steps.py]: remove commented lines

File: features/steps/project_steps.py [hunk 10]
```diff
@@ -1,25 +1,26 @@
...
+ #   logger.info("Status code of projects %s" % context.projects_response.status_code)
     context.list_of_projects = context.projects_response.json()
-    logger.info("List of projects %s" % context.list_of_projects)
+  #  logger.info("List of projects %s" % context.list_of_projects)
 
 
 @then(u'I validate project list is retrieved')
 def step_impl(context):
     logger.info(u'STEP: Then I validate project list is retrieved')
+    print("x")
```
Comment (d4vidmc) [inline features/steps/project_steps.py]: remove print

File: src/data/epic.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
+class Epic():
+    def __init__(self):
+        self.project_id = None  # int
+        self.name = None  # string
+        self.label = None  # string
+        self.label_id = None  # int
+        self.description = None  # string
```
Comment (d4vidmc) [inline src/data/epic.py]: Use dictionary template for epic variables

File: src/data/epic.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        self.branches = None  # List[branch]
+        self.followers = None  # List[follower]
+        self.follower_id = None  # List[int]
+        self.after_id = None  # int
+        self.before_id = None  # int
+
+    def get_body(self):
+        body = "{"
+        if self.name != None:
+            body = body + "\"name\":\"" + self.name + "\","
```
Comment (d4vidmc) [inline src/data/epic.py]: refactor this code

File: src/pivotal_services/base_service.py [hunk 13]
```diff
@@ -4,6 +4,14 @@
 
 class BaseService:
     def __init__(self):
+        # TODO: Add Property reader
+        # self.property_reader =
```
Comment (d4vidmc) [inline src/pivotal_services/base_service.py]: remove commented code

File: src/pivotal_services/epic_service.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
+from src.data.epic import Epic
+from src.pivotal_services.base_service import BaseService
+
+
+class EpicService(BaseService):
+
+    def __init__(self):
+        BaseService.__init__(self)
+        self.base_url = "https://www.pivotaltracker.com/services/v5"
```
Comment (d4vidmc) [inline src/pivotal_services/epic_service.py]: use config

File: src/pivotal_services/epic_service.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+    def get_url(self, url):
+        return self.base_url + url
+
+    def get_all_epics(self, project_id):
+        url = self.get_url("/projects/" + str(project_id) + "/epics")
+        return self.request_handler.get_request(url)
+
+    def get_all_epics_with_filter(self, project_id, filter):
+        # "https://www.pivotaltracker.com/services/v5/projects/$PROJECT_ID/epics?fields=id%2Ccomments"
```
Comment (d4vidmc) [inline src/pivotal_services/epic_service.py]: remove this

File: src/pivotal_services/epic_service.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+    def get_epic(self, epic_id):
+        url = self.get_url("/epics/" + epic_id)
+        return self.request_handler.get_request(url)
+
+    def create_epic(self, project_id, epic):
+        url = self.get_url("/projects/" + project_id + "/epics")
+        if isinstance(epic, Epic):
+            return self.request_handler.post_request(url, epic.getBody())
+        return self.request_handler.post_request(url, epic)
```
Comment (d4vidmc) [inline src/pivotal_services/epic_service.py]: change to
return self.request_handler.post_request(url, epic.get_body() if  isinstance(epic, Epic) else epic)

File: src/tests/epic_service_tests.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
+import pprint
+import unittest
+
+from src.pivotal_services.epic_service import EpicService
+
+project_id = 2322921
+epic_id=4275384
```
Comment (d4vidmc) [inline src/tests/epic_service_tests.py]: delete hardcoded variables

File: src/pivotal_services/epic_service.py [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+    def modify_epic(self, project_id, epic_id, epic):
+        url = self.get_url("/projects/" + project_id + "/epics/" + epic_id)
+        if isinstance(epic, Epic):
+            return self.request_handler.put_request(url, epic.getBody())
```
Comment (d4vidmc) [inline src/pivotal_services/epic_service.py]: use ternary condition

**Review by d4vidmc — CHANGES_REQUESTED** [review]

File: config.yml (hunk 1 above)
Comment (Fergitgirl) [reply config.yml]: please delete token

**Pull request merged**